Toggle revision history by clicking on last modified

gregwebs opened the pull request:
This is dissapointing since we only have metadata.
There is no way to link to a revision or show a diff.
The only thing that can be offered is an HTML download.

Doing a direct fetch of the HTML download runs into auth issues.
Perhaps there is a way to fix this, but I don't think it is very useful
to show old revisions without any diff.

There is an unofficial URL for getting a diff,
but it is a JSON output that needs rendering.
https://docs.google.com/document/d/{file.id}/showrevision?id={file.id}&end={rev.id}&start={rev.id}

breeswish commented on packages/website/src/pages/ContentPage/DocPage.tsx:
```diff
@@ -158,6 +192,35 @@ function DocPage({ file, renderStackOffset = 0 }: IDocPageProps) {
 
   return (
     <div style={{ maxWidth: '50rem' }}>
```
Since the rendering always take place in the `ContentPage` instead of globally, maybe you don't need the redux for storing such state. Just storing the state in the `ContentPage` should be enough :)

gregwebs replied:
okay, I was able to move it all to the FileAction view

breeswish commented:
The current UI works but does not looks good :) For example, it cannot collapse and the layout doesn't look nice.

![image](https://user-images.githubusercontent.com/1916485/124130684-9f343f00-dab1-11eb-91e0-676e2a203a4d.png)

I suggest that you can:
1. Use [Callout](https://developer.microsoft.com/en-us/fluentui#/controls/web/callout) to make it collapsible.
2. Use [Stack](https://developer.microsoft.com/en-us/fluentui#/controls/web/stack) to arrange your content with some paddings to make it looks nicer.

Thanks for the PR!

gregwebs commented:
It is already designed to be collapsible by clicking on the same button, that is a bug when I transitioned the state out of redux.
That is fixed now. I also adjusted the spacing and added an hr border for visual separation.

breeswish commented on packages/website/src/pages/FileAction.tsx:
```diff
@@ -26,10 +26,10 @@ function FileAction() {
   const { inMost: rInner, outMost: rOuter } = useRender();
-  const [revisionsEnabled, setRevisionsEnabled] = useState(false);
+  let [revisionsEnabled, setRevisionsEnabled] = useState(false);
   function toggleRevisions() {
```
It should be:

```js
const toggleRevisions = useCallback(() => {
    setRevisionsEnabled(v => !v);
}, []);

const commandBarItems: ICommandBarItemProps[] = useMemo(() => {
  ......
  onClick: () => { toggleRevisions(); },
}, [...., toggleRevisions])
```

gregwebs replied:
This doesn't work in my testing. That is, the modifications still show when I click on a new document.

breeswish replied:
I think they are different problems. You need to reset the `revisionsEnabled` to `false` when `rInner` or `rOuter` is changed. This is because the FileAction component is not unmounted when click on a new document.

gregwebs replied:
I pushed a commit which does this. It still doesn't seem like the right way, but like 2 other versions of this PR it works when tested.

breeswish commented on packages/website/src/pages/FileAction.tsx:
```diff
@@ -26,6 +26,13 @@ function FileAction() {
+    setRevisionsEnabled(revisionsEnabled);
+  }
+  const revs: Array<gapi.client.drive.Revision> = []
+  const [revisions, setRevisions] = useState(revs);
```
```suggestion
  const [revisions, setRevisions] = useStateArray<gapi.client.drive.Revision>([]);
```

gregwebs replied:
I don't see `useStateArray` anywhere.

breeswish replied:
My fault! It should be `useState<Array<gapi.client.drive.Revision>>([])`
